### ISSUE TITLE
Allow using iroh with other signers

### ISSUE THREAD

**Issue by flokli** [n0-computer/iroh#2355]
I'd like to use iroh with other signers, which are not a private `ed25519` key living in the host memory, but "external signer".

Some examples include:

 - FIDO tokens
 - Cloud HSM/Key vault things
 - Smartcards (PKCS#11 3.0)

We probably don't want to implement any of these in iroh itself (at least not for starters), but allow others to "bring their own signers".

Proposal:

 - make `iroh_net::key::SecretKey` a trait containing only `public()` and `sign()` (or using some of the traits from the `ed25519` crate)?
 - move the existing `SecretKey` to some `InMemorySecretKey` struct, which is constructed.
 - update the two consumers (`make_libp2p_extension` and `send_client_key` when I checked) to be generic (or use dynamic dispatch). It seems this is not in the path for each packet, so maybe dynamic dispatch is fine.

This will be an API change, but so might https://github.com/n0-computer/iroh/issues/1908 be too - so it might make sense to batch this together.

**Comment by matheus23** [comment]
There's two problems with this:
- I'm not sure we want to do an HSM access/Smartcard read or yubikey user interaction every time `SecretKey` is used in iroh for performance reasons.
- It'd need some thought on how such an API can be supported in an FFI context

An idea could be: we expose more utilities for making the `SecretKey` used in iroh more ephemeral by allowing to pass in self-signed certificate chains.
You'd then throw away the ephemeral key after shutdown and re-generate it on startup.
And you can sign the ephemeral key on startup outside iroh with a hardware-backed key and thus generate a self-signed certificate chain.

That said, this needs some thought in terms of threat model. If an online attacker can access the ephemeral key from memory and the certificates don't practically expire or aren't really revocable, it's as good as if we didn't have any certificate chains. If the certificate chains expire, this needs some extra thought on how to provide new certificates to iroh every time they do. If the certificates can be revoked, there needs to be some thought on how revocations are stored.

---

It'd be helpful to look into how other projects solve this. I know AWS's CloudHSM for example take over the whole TLS connection establishment.

**Comment by flokli** [comment]
> There's two problems with this:
> 
>     * I'm not sure we want to do an HSM access/Smartcard read or yubikey user interaction every time `SecretKey` is used in iroh for performance reasons.

I'd expect using these things for key storage would need to be configured explicitly, so it'd be some kind of opt-in. How bad is the performance penalty? I'd assume it'd only run when establishing a new connection, not after that?

>     * It'd need some thought on how such an API can be supported in an FFI context

:+1: 

> An idea could be: we expose more utilities for making the `SecretKey` used in iroh more ephemeral by allowing to pass in self-signed certificate chains. You'd then throw away the ephemeral key after shutdown and re-generate it on startup. And you can sign the ephemeral key on startup outside iroh with a hardware-backed key and thus generate a self-signed certificate chain.

This would be a very big change in architecture. Right now, the ed25519 key "becomes your identity", and cycling through these on every startup mandates solving the whole PKI-related questions you wrote.

The proposal in here already effectively prevents key material from being stolen/copied, by putting it into some smartcard/HSM, without change in protocol.

Rotating these keys, and using a PKI and certificates to establish which keys to trust is something some setups might want, and can be done on top of it, at a lower cadence (so maybe it's fine if the identities change) - but generally I see that an orthogonal issue.

**Comment by matheus23** [comment]
> This would be a very big change in architecture. Right now, the ed25519 key "becomes your identity", and cycling through these on every startup mandates solving the whole PKI-related questions you wrote.

Yes. The idea would include rooting your identity in the root of the certificate chain instead of the ephemeral key.

**Comment by flokli** [comment]
I feel these two things are a bit orthogonal to each other. I can understand you don't want to deal with this issue before that's being implemented, but there should probably be a separate tracking issue for the whole identity revamp.

**Comment by matheus23** [comment]
I don't think these issues are orthogonal: If we do an "identity revamp" we won't need to introduce a trait for `SecretKey` in the iroh codebase.
Perhaps I misunderstand your use-case though. What's your end goal? Why do you want to move the key out of working memory? Does your threat model include that a malicious actor with working memory access? Do you want to be able to recover from such a compromise or do you want to protect from such a compromise completely?

**Comment by flokli** [comment]
I don't want to have key material on disk, primarily. Which is why it'd be nice for these to be living in the TPM, bound to some PCRs, or using FIDO/PIV.

**Comment by rklaehn** [comment]
@flokli this is not going to work for all things, but I think we are going to attempt making at least the signing more flexible. I have recently played around with FROST, see https://github.com/rklaehn/frosty . There will be a blog post in the near future about this as well.

We are probably going to keep PrivateKey put pass the discovery traits a "signer" instead of always taking a private key. This is to support the use case where you don't have the private key but just "a means to sign".

**Comment by Torhamilton** [comment]
Hi Guys,
I was really hoping to use a signer like [lair-keystore](https://crates.io/crates/lair_keystore_api). My use case, requires private keys away from main app. 
Thanks

**Comment by lorenzleutgeb** [comment]
Users of Radicle are asking for this, too. I arrived at the same conclusion as @flokli.

You could build on top of [`signature::Signer`](https://docs.rs/signature/3.0.0/signature/trait.Signer.html) for [`ed25519::Signature`](https://docs.rs/ed25519/3.0.0/ed25519/struct.Signature.html). See [`impl signature::Signer<ed25519::Signature> for ed25519_dalek::SigningKey`](https://docs.rs/ed25519-dalek/3.0.0/ed25519_dalek/struct.SigningKey.html#impl-Signer%3CSignature%3E-for-SigningKey) and [`impl signature::Signer<ed25519::Signer> for yubihsm::Signer`](https://docs.rs/yubihsm/0.42.1/yubihsm/ed25519/struct.Signer.html) (albeit for an older version of `ed25519`).

For Radicle we use
```rust
trait Signer
where
    Self: signature::Signer<Signature>,
    Self: signature::Keypair<VerifyingKey = VerifyingKey>,
    Self: AsRef<PublicKey>,
{
    /// Return a reference to the [`PublicKey`].
    ///
    /// This is generally satisfied by the [`AsRef<PublicKey>`] instance.
    fn public_key(&self) -> &PublicKey {
        self.as_ref()
    }
}
```
which serves a similar purpose and allows us to abstract over a key in memory (loaded from the filesystem) and using [`ssh-agent`](https://en.wikipedia.org/wiki/Ssh-agent).
